### PR TITLE
Polish

### PR DESCRIPTION
**Pull request by DavidVorick**
The documentation should be accurate now, this wasn't the case before this PR. Also, a lot of the logic was cleaned up to use simpler constructions. Some more tests were added, including a few more edge cases I found reason to worry about (no bugs revealed), and some tests to check the consistency of proofs on much larger Merkle trees.

Despite not having mutation testing, I feel quite confident in this package's correctness. I've been thorough, and tried hard to keep the extensiveness of the test suite greater than the complexity of the codebase.

### CODE REVIEW

File: merkletree.go [hunk 1]
```diff
@@ -13,33 +14,38 @@ import (
 // A Tree takes data as leaves and returns the merkle root. Each call to 'Push'
 // adds one leaf to the merkle tree. Calling 'Root' returns the Merkle root.
 // The Tree also constructs proof that a single leaf is a part of the tree. The
-// leaf can be chosen with 'SetIndex'.
+// leaf can be chosen with 'SetIndex'. The memory footprint of Tree grows in
+// O(log(n)) in the number of leaves.
 type Tree struct {
+	// The Tree is stored as a stack of subtrees. Each subtree has a height,
+	// and is the Merkle root of 2^height) leaves. A Tree with 11 nodes is
```
Comment (lukechampine) [inline merkletree.go]: typo

File: merkletree.go [hunk 2]
```diff
@@ -13,33 +14,38 @@ import (
...
 
-// A subTree is a subTree in the Tree. 'height' refers to how tall the subTree
-// is. The children of the tree are not accessible, as they have already been
-// hashed into 'sum'. 'next' is the next subTree, and is guaranteed to have
-// a higher height unless it is nil.
-
-// A subTree contains the merkle root of a complete (2^n leaves) subTree of
-// the Tree. 'sum' is the Merkle root of the subTree. If 'next' is not nil, it
-// will be a tree with a higher height.
+// A subTree contains the merkle root of a complete (2^height leaves) subTree
```
Comment (lukechampine) [inline merkletree.go]: Merkle

File: merkletree.go [hunk 3]
```diff
@@ -51,12 +57,38 @@ func sum(h hash.Hash, data []byte) []byte {
...
+//		Hash( 0x01 || left sibling sum || right sibling sum)
+func nodeSum(h hash.Hash, a, b []byte) []byte {
 	return sum(h, append(append([]byte{1}, a...), b...))
 }
 
+// joinSubTrees combines two equal sized subTrees into a larger subTree.
+func (t *Tree) joinSubTrees(a, b *subTree) *subTree {
+	if DEBUG {
+		if b.next != a {
+			panic("invalid subtree join - 'a' is not paried with 'b'")
```
Comment (lukechampine) [inline merkletree.go]: paired

File: README.md [hunk 4]
```diff
@@ -9,8 +9,8 @@ BitTorrent, for example, requires downloading many small pieces of a file from
 many untrusted peers; Merkle proofs allow the downloader to verify that each
 piece is part of the full file.
 
-When sha256 is used as the hashing algorithm, this merkle tree will match the
-merkle tree outlined in RFC 6962, 'Certificate Transparency'.
+When sha256 is used as the hashing algorithm, the implementation matches the
+merkle tree described in RFC 6962, 'Certificate Transparency'.
```
Comment (lukechampine) [inline README.md]: Merkle

File: merkletree.go [hunk 5]
```diff
@@ -13,33 +14,38 @@ import (
...
+	// tree. When a new leaf is inserted, it is inserted as a subtree of height
+	// 0. If there is another subtree of the same height, both can be removed,
+	// combined, and then inserted as a subtree of height n + 1.
 	head *subTree
 	hash hash.Hash
 
-	// Variables to help build proofs that the data at 'proofIndex' is in the
-	// merkle tree.
+	// Helper variables used to construct proofs that the data at 'proofIndex'
+	// is in the merkle tree. The proofSet is constructed as elements are being
```
Comment (lukechampine) [inline merkletree.go]: Merkle

File: merkletree.go [hunk 6]
```diff
@@ -230,96 +267,119 @@ func (t *Tree) Prove() (merkleRoot []byte, proofSet [][]byte, proofIndex uint64,
 
 // VerifyProof takes a Merkle root, a proofSet, and a proofIndex and returns
 // true if the first element of the proof set is a leaf of data in the Merkle
-// root.
+// root. False is returned if the proof set or merkle root is nil, and if
```
Comment (lukechampine) [inline merkletree.go]: Merkle

File: merkletree.go [hunk 7]
```diff
@@ -230,96 +267,119 @@ func (t *Tree) Prove() (merkleRoot []byte, proofSet [][]byte, proofIndex uint64,
...
+// 'numLeaves' equals 0.
 func VerifyProof(h hash.Hash, merkleRoot []byte, proofSet [][]byte, proofIndex uint64, numLeaves uint64) bool {
-	if len(proofSet) == 0 || merkleRoot == nil || numLeaves == 0 {
+	// Return false for nonsense input.
+	switch {
+	case merkleRoot == nil:
+		return false
+	case numLeaves == 0:
```
Comment (lukechampine) [inline merkletree.go]: A note here (and elsewhere) about why a `switch` is used would be nice.

File: merkletree.go [hunk 8]
```diff
@@ -230,96 +267,119 @@ func (t *Tree) Prove() (merkleRoot []byte, proofSet [][]byte, proofIndex uint64,
...
 		return false
 	}
 
-	// The first element of the proof set is the original data. Prepend it with
-	// 0x00 and hash it to get the first level subTree root.
+	// In a merkle tree, every node except the root node has a sibling.
```
Comment (lukechampine) [inline merkletree.go]: Merkle
